src/conversions.jl

```diff
@@ -0,0 +1,302 @@
+function TriMesh(p::PointCloud, res::Int=32; algo=:MarchingCubes)
+    voxel = pointcloud_to_voxel(p, res)
+    v = VoxelGrid(voxel)
+    return TriMesh(v; algo=algo)
+end
+
+function TriMesh(v::VoxelGrid; thresh::Number=0.5f0, algo=:MarchingCubes)
+    verts,faces = voxel_to_trimesh(v,Float32(thresh),algo)
+    return TriMesh(verts,faces)
+end
+
+function PointCloud(m::TriMesh, npoints::Int=1000)
+    p = sample_points(m, npoints)
+    return PointCloud(p)
+end
+
+function PointCloud(v::VoxelGrid, npoints::Int=1000; thresh::Number=0.5f0, algo=:MarchingCubes)
+    m = TriMesh(v, thresh=thresh, algo=algo)
+    points = sample_points(m, npoints)
+    return PointCloud(points)
+end
+
+function VoxelGrid(m::TriMesh,res::Int=32)
+    v = trimesh_to_voxel(m,res)
+    return VoxelGrid(v)
+end
+
+function VoxelGrid(p::PointCloud, res::Int=32)
+    v = pointcloud_to_voxel(p, res)
+    return VoxelGrid(v)
+end
+
+function pointcloud_to_voxel(
+    pcloud::PointCloud,
+    resolution::Int = 32,
+)
+    p = cpu(pcloud.points)
+    _, _N, _B = size(p)
+    verts_max = maximum(maximum(p, dims = 1), dims = 2)
+    verts_min = minimum(minimum(p, dims = 1), dims = 2)
+    cloud = ((p .- verts_min) ./ (verts_max - verts_min))
+
+    ind1 = Int[]
+    ind2 = Int[]
+    ind3 = Int[]
+
+    for x = 1:resolution, y = 1:resolution, z = 1:resolution
+        push!(ind1, x)
+        push!(ind2, y)
+        push!(ind3, z)
+    end
+    ind1 = reshape(ind1, 1, :)
+    ind2 = reshape(ind2, 1, :)
+    ind3 = reshape(ind3, 1, :)
+
+    grid_points = cat(ind1, ind2, ind3, dims = 1)
+    grid_points = ((grid_points .+ 0.5) ./ resolution)
+
+    nn_idx = cat(
+        [
+            CartesianIndex.(
+                reduce(vcat, knn(KDTree(cloud[:, :, i]), grid_points, 1)[1]),
+                i,
+            ) for i = 1:_B
+        ]...,
+        dims = 2,
+    )
+
+    dists_vec = (grid_points .- cloud[:, nn_idx])
+    dists = sum((dists_vec .^ 2), dims = 1)
+    dists = reshape(dists, resolution, resolution, resolution, _B)
+    voxels = typeof(similar(pcloud.points,1,1,1,1))(dists .<= (0.6 / (resolution * resolution)))
+    return voxels
+end
+
+function trimesh_to_voxel(m::TriMesh{T,R,S},res::Int=32)    where {T,R,S}
+    verts_list = get_verts_list(m)
+    faces_list = get_faces_list(m)
+
+    _B = length(verts_list)
+    voxels = S{T,4}(undef,res,res,res,_B)
+    # print(typeof(verts_list))
+    for (i,v,f) in zip(1:_B,verts_list, faces_list)
+        voxels[:,:,:,i] = _voxelize(cpu(v),f,res)
+    end
+    return voxels
+end
+
+function _voxelize(
+    v::Array{<:AbstractFloat,2},
```

Comment:
Does it need to be abstract float here, what about real?

Reply:
verts in TriMesh is restricted to AbstractFloat, therefore I only included the same here, although changing to Real won't be any problem